lib/datadog/ci/ext/environment.rb

```diff
@@ -312,6 +317,107 @@ def extract_bitrise(env)
           }
         end
 
+        def git_commit_users
+          # Get committer and author information in one command.
+          #
+          # Because we can't get a reliable UTC time from all recent versions of git
+          # We have to rely on converting the date to UTC ourselves.
+          output = exec_git_command("git show -s --format='%an\t%ae\t%at\t%cn\t%ce\t%ct'")
+          return unless output
+
+          fields = output.split("\t").each(&:strip!)
+
+          author = {
+            name: fields[0],
+            email: fields[1],
+            date: Time.at(fields[2].to_i).utc.to_datetime.iso8601
+          }
+
+          committer = {
+            name: fields[3],
+            email: fields[4],
+            date: Time.at(fields[5].to_i).utc.to_datetime.iso8601
+          }
+
+          [author, committer]
+        rescue => e
+          Datadog.logger.debug("Unable to read git commit users: #{e.message} at #{e.backtrace.first}")
+          nil
+        end
+
+        def git_repository_url
+          exec_git_command('git ls-remote --get-url')
+        rescue => e
+          Datadog.logger.debug("Unable to read git repository url: #{e.message} at #{e.backtrace.first}")
+          nil
+        end
+
+        def git_commit_message
+          exec_git_command('git show -s --format=%s')
+        rescue => e
+          Datadog.logger.debug("Unable to read git commit message: #{e.message} at #{e.backtrace.first}")
+          nil
+        end
+
+        def git_branch
+          exec_git_command('git rev-parse --abbrev-ref HEAD')
+        rescue => e
+          Datadog.logger.debug("Unable to read git branch: #{e.message} at #{e.backtrace.first}")
+          nil
+        end
+
+        def git_commit_sha
+          exec_git_command('git rev-parse HEAD')
+        rescue => e
+          Datadog.logger.debug("Unable to read git commit SHA: #{e.message} at #{e.backtrace.first}")
+          nil
+        end
+
+        def git_tag
+          exec_git_command('git tag --points-at HEAD')
+        rescue => e
+          Datadog.logger.debug("Unable to read git commit SHA: #{e.message} at #{e.backtrace.first}")
+          nil
+        end
+
+        def git_base_directory
+          git_dir = exec_git_command('git rev-parse --git-dir')
+          return unless git_dir
+
+          # The result will point to the '.git' directory.
+          # We actually want the base source directory.
+          File.expand_path('..', git_dir)
```

Comment:
Not sure if we want to care about it, but note that this is not correct in the presence of git worktrees.

I keep a single checkout of dd-trace-rb but multiple worktrees (to be able to keep several branches open at once). For those, this doesn't look correct:

```bash
ivo.anjo@macieira:~/datadog/dd-trace-rb$ cat ../base-directory.rb
puts `git rev-parse --git-dir`.strip
puts File.expand_path('..', `git rev-parse --git-dir`.strip)
ivo.anjo@macieira:~/datadog/dd-trace-rb$ ruby ../base-directory.rb
.git
/Users/ivo.anjo/datadog/dd-trace-rb
ivo.anjo@macieira:~/datadog/dd-trace-rb$ cd ../second-dd-trace-rb/
ivo.anjo@macieira:~/datadog/second-dd-trace-rb$ ruby ../base-directory.rb
/Users/ivo.anjo/datadog/dd-trace-rb/.git/worktrees/second-dd-trace-rb
/Users/ivo.anjo/datadog/dd-trace-rb/.git/worktrees
```

Not we may not want to care about this for CI, but just thought I'd point out (since I was curious myself)

Reply:
I changed it so that it works with worktrees now.
Using `git rev-parse --show-toplevel` seems more correct in any case.

I tried to add a test case for worktrees using the `GIT_DIR` environment variable, like I'm using for git testing here, but couldn't get it to work.

I tried different combinations of environment variables, but couldn't get: a worktree that's a spin-off of `GIT_DIR=spec/datadog/ci/ext/fixtures/git/gitdir_with_commit` that works correctly inside a spec running inside a `dd-trace-rb` directory.

For now, I'm ok with this case, and I like that `git rev-parse --show-toplevel` doesn't perform any `../` path shenanigans.